From a0d5ecb589895f208a0147d5e1ccba569d3f8482 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?La=C3=A9rcio=20de=20Sousa?= <lbsousajr@gmail.com>
Date: Wed, 6 Nov 2013 12:55:26 -0200
Subject: [PATCH] Delay non-seat0 display servers startup

This workaround can be used to avoid a race condition
between seat0 and non-seat0 display servers, ensuring
seat0 display server will start first.

If a non-seat0 display server starts before the seat0 one,
that one can "steal" the VT expected by this one. It may result
in inactive systemd-logind graphical sessions for seat0.
---
 daemon/gdm-server.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/daemon/gdm-server.c b/daemon/gdm-server.c
index 91a04e6..ef602f8 100644
--- a/daemon/gdm-server.c
+++ b/daemon/gdm-server.c
@@ -749,6 +749,15 @@ gdm_server_start (GdmServer *server)
 #endif
         }
 
+#ifdef WITH_SYSTEMD
+        /* Delay non-seat0 display servers startup,
+         * in order to avoid a race condition that
+         * may result in inactive logind sessions for seat0. */
+        if (LOGIND_RUNNING() && (strcmp (server->priv->display_seat_id, "seat0") != 0)) {
+                sleep (1);
+        }
+#endif
+
         /* fork X server process */
         if (!gdm_server_spawn (server, vtarg, error)) {
                 goto out;
-- 
1.8.3.1

